open_seq2seq/data/data_layer.py

```diff
@@ -121,6 +121,15 @@ def input_tensors(self):
     """
     pass
 
+  def build_interactive_graph(self):
```

Comment:
Is it necessary to create tf.data objects in the implementation of this function? I think it will be better and clearer to just define all necessary placeholders as self.placeholders or something and require user to pass their values (or even dictionary -- placeholder: value) to the function that will query the model for the output (see my other comment about this). It will also save the users from copy/pasting a lot of code from the main build_graph function. The only problem with this approach is that there might be some functionality inside the tf.data object that we want to duplicate on the user input, such as image/audio proprocessing etc. But most of the time this functionality is defined as a Python function or a chain of TF operations, both of which can be applied without invoking tf.data either on the input or on the placeholders directly before feeding them into the network. Maybe it will be better to isolate this function into some separate abstraction (like another method of the DataLayer class?) and call it on the placeholders or input and have the same functions passed as pyfunc to tf.data object. I'm not sure if it will be easy to do for general preprocessing, but it seems to me like a better way than to create the whole new tf.data object for processing of a single sample.

Reply:
I agree that creating a new tf.data object for processing a single sample is overkill. Now that I think on it more, it makes more sense to have self._input_tensors just be a placeholder that accepts what we need for each model. 

In this case, we would need to define two functions. One that creates the placeholders for self._input_tensors so we can use it during model.compile() to build the graph. Then a second function that preprocesses an input and creates the necessary feed_dict.